### ISSUE TITLE
Optimize Equihash implementation

### ISSUE THREAD

**Issue by nathan-at-least** [zcash/zcash#857]
There are several ways our initial release of Equihash are suboptimal for memory usage and time. Let's do a subsequent release to improve those so that we can set the parameters (#856) appropriately.

We don't have to do every possible optimization, but enough that it's feasible to run a miner with the right parameters.

**Comment by str4d** [comment]
Optimisations I have working:

- Truncate the indices (option 2 in the paper).
- Store the truncated indices in the same `unsigned char*` as the hash.
- Initialise `Equihash` using template parameters.
  - Enables all Equihash parameters to be validated at compile time.

Proposed optimisations:

- Calculate the hash on-the-fly (option 1 in the paper).
- Use compile-time values for hash lengths and number of indices.
  - Removes two 4-byte length fields per row at the expense of either complicated templating code, or virtual functions which add a 4-byte pointer to a VMT per row.
- Initialise the `unsigned char*` as a fixed-width `unsigned char[]`
  - Removes a 4-byte pointer per row at the expense of memory usage per row equal to the peak round requirements.
  - Additionally, has same expenses as compile-time values optimisation.
- Replace usage of `std::vector` with plain arrays
  - Reduces memory usage at the expense of complexity, performing our own memory management.

**Comment by str4d** [comment]
At this point, both optimisations from the paper to reduce peak memory usage have been implemented. #921 contains the index truncation optimisation as well as fixed-length rows, and the XORing on-the-fly optimisation is implemented in a separate branch while performance issues are sorted out.

I haven't managed to get a massif run to complete yet, but using regtest mode as an indicator, reading the RES column in htop and conservatively assuming 22MB for general zcashd overhead:

`n` | `k` | Revision | Optimisations | Observed | Paper | Overhead
--- | --- | --- | --- | --- | --- | --- | ---
96 | 3 | e587472 | Fixed-width, 8-bit indices | 359MB | 320MB | 39MB
96 | 3 | 71468ad | Fixed-width, 8-bit indices, XOR on-the-fly | 342MB | 320MB | 22MB
144 | 5 | e587472 | Fixed-width, 8-bit indices | 747MB | 704MB | 43MB
144 | 5 | 71468ad | Fixed-width, 8-bit indices, XOR on-the-fly | 727MB | 704MB | 23MB

The observed memory usage is from the first sorting step; the number of elements in the list doesn't change much over the rounds, and because of the fixed-width rows that means the initial round usage is the roughly the same as the final round usage. However, actual peak memory might be higher during the colliding stage. I'm not exactly sure why the XORing results are slightly lower than the non-XORing ones, because the number of stored rows should be the same and the row sizes are unchanged.

Referring to my earlier list, all that remains is replacing the remaining uses of `std::vector` (which is possibly the source of the remaining overhead), and rewriting the second half of the optimised solver to only keep up to `k` lists in memory instead of up to `2^k`.

**Comment by str4d** [comment]
It should also be noted that e587472 crashes for `n = 96, k = 3`, because the initial row length (full hash + 1 truncated index) is too long for the final row size (`2n/(k+1)`-bit hash + `2^(k-1)` truncated indices). This is not encountered with on-the-fly XORing because then the peak row length is always in the last step (per the paper).

**Comment by daira** [comment]
This is excellent! We should now be able to use the final parameters (probably n = 144, k = 5) for the Protocol 2016.0-beta release :-)

**Comment by str4d** [comment]
Here is local timing data for `n = 144, k = 5` on e587472 (corresponding to 747MB memory usage above):

```json
[
    {
        "runningtime" : 815.47173900
    },
    {
        "runningtime" : 2335.82262100
    },
    {
        "runningtime" : 345.43088500
    },
    {
        "runningtime" : 730.35247400
    },
    {
        "runningtime" : 621.40392300
    },
    {
        "runningtime" : 1413.99249400
    },
    {
        "runningtime" : 761.86925800
    },
    {
        "runningtime" : 589.17610100
    },
    {
        "runningtime" : 502.41565300
    },
    {
        "runningtime" : 345.59338300
    }
]
```

The wide variance in runtime stems directly from the number of partial truncated-index solutions that need to be sifted through (sometimes a few dozen, sometimes a few hundred). The number of partial solutions drops as more rounds are done with the XOR on-the-fly optimisation, which trades off culling runtime for XORing. However, even the fastest runs above were still over five minutes, which indicates there is inherent optimisation that can also be done.

The next step for this ticket is to benchmark the individual components of the optimised solver, and determine what operations or structures are slowing the solver down.

**Comment by str4d** [comment]
The above benchmarks were measured on an Intel Core i5-3570K @ 3.40GHz, however note that I was running them on Xubuntu in VirtualBox on Windows 7.

**Comment by zookozcash** [comment]
Learning to use a profiling tool might be more efficient than benchmarking subcomponents by hand.

**Comment by str4d** [comment]
#988 implements depth-first scanning to reduce variability during the elimination of partial solutions.

Other possible speed optimisations:

- [ ] XOR on-the-fly
  - Reduces variability in solving time at the expense of requiring additional time during the main step, so may not be worthwhile past the first round.
- [ ] Don't truncate indices for some rounds, but still store full hash
  - May not be possible, as this only works if the full row lengths in those rounds are below the peak truncated row length for the chosen parameters.
- [ ] Multithreaded solver
  - [ ] Multithreaded sorting
  - [ ] Eliminate partial solutions in parallel
  - Having the solver use more threads will reduce the number of mining threads that can be run (`genproclimit`).
  - See comments [here](https://github.com/zcash/zcash/pull/988#issuecomment-223893286) and [here](https://github.com/zcash/zcash/pull/988#issuecomment-224216718).

**Comment by daira** [comment]
Another possible optimization is to change the sorting algorithm to bucket sort, proxmap sort, or similar. This should in principle be more efficient because the elements being sorted are expected to be approximately uniformly distributed.

**Comment by solardiz** [comment]
Regarding multi-threading the solver: let's first recall that with Equihash we appear to be relying mostly on the memory bandwidth as the limiting factor for ASIC miners. Not on memory size per instance, which an ASIC will amortize anyway due to the parallelism within one instance (the feasibility of such "attack" will need to be confirmed by implementing the full Equihash with our parameters perhaps on GPU first, but currently it seems likely). Also, let's recall that our typical target devices are at least quad-core. Thus, maybe it makes sense to reduce the per-instance memory to 1/4 of our target memory usage on our target devices, and just use threads at the higher level where we reasonably can? And this will be almost everywhere, since devices with more than 4 cores will also tend to have more memory. This approach should partially mitigate #1055 by reducing the time needed to compute one solution from ~50 to ~13 seconds on a fast desktop CPU (and to less than that with other, unrelated optimizations).

Multi-threading the solver now will be valuable from standpoint of evaluation of GPU mining and FPGA/ASIC mining ("attack") potential, which demand that much of whatever parallelism is available actually be exploited. However, it possibly isn't as valuable from CPU mining perspective (and is a premature optimization, which we would then need to include in the desirable code refactoring from C++ to C, and merge with whatever revisions to Equihash we might be making) unless our memory requirements are such that e.g. smartphone mining utilizing all CPU cores (and all memory bandwidth, which is usually _not_ available from just one thread) isn't otherwise possible. So maybe we need to reconsider our memory requirements per instance, so that 4 instances fit in our current target smartphones.

I am not convinced the choice of Equihash is right, and I am skeptical about smartphone mining, but within those constraints and with my current (way too limited and possibly incorrect) knowledge of possible Equihash optimizations/parallelization, I arrive at the reasoning above. So I thought I'd share.

**Comment by jl777** [comment]
I am offering a 1 BTC bounty for "which we would then need to include in the desirable code refactoring from C++ to C"

it is a required step to achieving a GPU miner, but also will help simple C programmers like me

**Comment by zookozcash** [comment]
I think @solardiz's question/suggestion about multithreading is spot on. The question of whether to implement multithreading for a single instance or instead to have multiple smaller instances should be answered, based on our strategic goals for PoW, before we implement multithreading for a single instance.

**Comment by Cryptoslave** [comment]
I am also not convinced the choice of Equihash is right, and I am also skeptical about smartphone mining, my nexus 5 becomes empty while plugged in even just browsing.... ;)

About the first point, I think it would be interesting if you consider again Argon2 instead of Equihash especially since running time is fast and verification is quick. I still do not see why Equihash was favored over Argon2, both from khovratovich. 

The author said for Equihash: "a reference implementation of a proof-of-work requiring 700 MB of RAM runs in 30 seconds on a 1.8 GHz CPU"... but instant to verify.

And the author said about Argon2:
"Cryptocurrency mining, that takes 0.1 seconds on a 2 Ghz CPU using 1 core - Argon2d with 2 lanes and 250 MB of RAM;
Backend server authentication, that takes 0.5 seconds on a 2 GHz CPU using 4 cores - Argon2d with 8 lanes and 4 GB of RAM."
"Performance. One-pass Argon2 is 50% faster than scrypt. Two-pass, two-thread Argon2 is as fast as scrypt".
"To strengthen the resistance to tradeoff attacks, we propose to make 3 passes over memory. On our 1.8 GHz machine this scheme runs in 2 seconds."
"The ranking method with reduction by 2 applied to 1-lane, 3-pass Argon2d yields the average depth of the recomputation tree is 20.7, thus the time penalty is 21.7. So we conclude that the time-memory product will grow at least by the factor of 7, and the same holds for the time-area product." 

also verify quickly:
"Finally, we offer a concrete proof-of-work function called Argon2 based on the data-dependent version Argon2d, which is offered for ASIC- and botnet-resistant cryptocurrencies. It requires 2GB of memory for generation but only 1390KB for verification, and can be computed in 2 seconds on a 2 GHz PC, while its verifcation takes milliseconds."

Argon2 also won the Password Hashing Competition so we can assume it was well reviewed by peers.

So 30 sec running time for Equihash or 2 sec for best settings on Argon2, with both being anti botnet memory hard hashing algorithm for cryptocurrencies by the same authors, I do not understand why Equihash was favored really. There would not be so many issues with mining Argon2 I guess.

http://eprint.iacr.org/2015/430.pdf
https://www.cryptolux.org/images/0/06/Argon2-euro.pdf

**Comment by str4d** [comment]
@Cryptoslave at a security level where miners cannot feasibly forge the proof of work, Argon2 requires 2MB of data in the block header to validate it. Given we want to use 2MB blocks, we'd either have no space for transactions, or would need to actually use 4MB blocks.

**Comment by tjade273** [comment]
@cryptoslave Equihash is far more practical on Ethereum than Argon. 1 MB of memory is essentially impossible, it would be incredibly expensive to process. Equihash seems to use on the order of 1KB of memory, if I'm not mistaken. 

So if you want Zcash to be amenable to an Ethereum relay, verification memory usage needs to be kept as low as possible.

**Comment by Cryptoslave** [comment]
@str4d  @tjade273 Thank you both I understand now :+1: 
Well this is an excellent argument ;)
Time to tune equihash more then...

**Comment by solardiz** [comment]
> reduce the per-instance memory to 1/4 [...] This approach should partially mitigate #1055 by reducing the time needed to compute one solution from ~50 to ~13 seconds

Correction: the expected speedup is somewhat greater than 4x, and that's great. I keep forgetting Equihash isn't in the same camp with traditional memory-hard PoWs.

**Comment by solardiz** [comment]
> It requires 2GB of memory for generation but only 1390KB for verification

> Argon2 requires 2MB of data in the block header to validate it

... but we don't have to use 2 GB for generation - we can choose less, especially if we're comparing vs. Equihash tuned to 1/4 the current target size. This will reduce the memory needs for verification and how much data to store in the block header as well, right?

**Comment by str4d** [comment]
By my reading of section 8.2, the number of blocks that must be opened for a particular security level is independent of the target memory usage. That is, we'd need to open ~500 blocks even if we targeted <1GB, so without changing the block the proof size will be the same.

**Comment by solardiz** [comment]
This should probably be discussed on another GitHub issue (which one?) or elsewhere, but just to comment on what was already said in here:

> So 30 sec running time for Equihash or 2 sec for best settings on Argon2, with both being anti botnet memory hard hashing algorithm for cryptocurrencies by the same authors

I don't see anything anti-botnet in either. Out of the 22 non-withdrawn PHC entries and Equihash, the only ones with anti-botnet capabilities were EARWORM and yescrypt, and only with ROM (lookup table) size in excess of typical botnet nodes' RAM size. And EARWORM, lacking a RAM portion, could be efficiently attacked on a botnet anyway, through buffering/grouping of lookups from a large number of concurrent instances. So this leaves only yescrypt. (Disclosure: I designed yescrypt.)

And this reminds me: with Equihash containing plenty of parallelism anyway, why don't we go for a more traditional memory-hard PoW combined with a ROM lookup table? For example, 16 MB RAM + 1 GB ROM (this is not enough for anti-botnet, but that's not my point; if we want a botnet-resistant coin, it would also be unusable on smartphones). We could use yescrypt, or we could enhance Argon2 with a ROM, and either way we'll get quick and small verification (except that it would need the ROM, but I think that's OK). I think this would effectively provide the same properties that we want from Equihash while also providing far greater confidence (to me, at least) on what optimizations and speedups are (not) practical on what devices.

**Comment by solardiz** [comment]
FWIW, running @str4d's standalone miner and interrupting/continuing it while in gdb (poor man's profiling suitable for low-hanging fruit), I see it spends much time in: blake2b_compress_ref(), memcpy(), memcmp(), and memmove(). #1049 might deal with memcmp(), but we also need to deal with the rest of these four: switch to SIMD-enabled implementation of BLAKE2b (*_ref() means the non-SIMD reference implementation, right?) and look into avoiding memory copying/moving in favor of doing those operations for just the pointers/indices (or are those calls being invoked on indices only already? if so, no need to use the libc functions there - it should be just a few instructions).

**Comment by jl777** [comment]
any chance of a C implementation? at least for some parts of it. I can then work on optimizing it. Any GPU version will need to start from a C implementation

**Comment by solardiz** [comment]
@jl777 Yes. Earlier today, I talked to a friend about him giving this a try. If not him, I might try to find time for it myself. I am also trying to get @zookozcash's OK/understanding/approval for this.

**Comment by jl777** [comment]
I am offering a 1 BTC bounty to help spur this along

**Comment by str4d** [comment]
If a C implementation is likely to be where we need to go to get the needed optimisations (rather than continuing with the existing C++ impl), then I'm happy to help work on it (subject to approval by @nathan-at-least).

**Comment by solardiz** [comment]
@str4d Thanks. I expect that a C implementation will be needed to come up with some of the optimizations, and as a base for porting to OpenCL, to CUDA, and to devices that easily have C, but not C++/Boost. It might happen that we would later be able to merge the same optimizations into the C++ implementation and get it to run almost as fast, but without a pure C baseline this is impossible to know. Then we'll need to decide which implementation to keep as primary in the tree. I think it'll be simpler to go pure C for everything, but we might have options.

**Comment by jl777** [comment]
C will always be a little bit to a lot faster, depending on specific implementations. The fewer dependencies, the more portable, but OpenCL and CUDA are C subsets and not C supersets, so if GPU implementations are wanted, then definitely C is the way to go. there are already blake2b implementations in C

**Comment by daira** [comment]
libsodium is C (and optional asm). No need to complicate things by changing more than we need to.

**Comment by solardiz** [comment]
> interrupting/continuing it while in gdb (poor man's profiling suitable for low-hanging fruit)

FWIW, trying to compile with `-pg` for not-so-poor man's profiling, I get failure at linking:

```
/opt/rh/devtoolset-3/root/usr/libexec/gcc/x86_64-redhat-linux/4.9.1/ld: /usr/lib/../lib64/gcrt1.o: relocation R_X86_64_32S against `__libc_csu_fini' can not be used when making a shared object; recompile with -fPIC
```

Looks like it's this bug: https://sourceware.org/bugzilla/show_bug.cgi?id=17836

Dropping `-pie` from LDFLAGS, I got most things to compile and link successfully, only failing at:

```
  CXXLD    zcash-gtest
/opt/rh/devtoolset-3/root/usr/libexec/gcc/x86_64-redhat-linux/4.9.1/ld: .libs/zcash-gtest: hidden symbol `_ZN5boost6system15system_categoryEv' in /home/z/standalone-miner-build/depends/x86_64-unknown-linux-gnu/lib/libboost_system-mt.a(error_code.o) is referenced by DSO
/opt/rh/devtoolset-3/root/usr/libexec/gcc/x86_64-redhat-linux/4.9.1/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
```

but I don't care for now - I got standaloneminer and zcashd (and more), and that's all I wanted to profile anyway - will try now.

**Comment by daira** [comment]
For the ld error, see http://stackoverflow.com/questions/23696585/what-does-exactly-the-warning-mean-about-hidden-symbol-being-referenced-by-dso .

**Comment by solardiz** [comment]
Here's a profiling run:

```
[z@super src]$ time ./standaloneminer
Running Equihash solver with nNonce = 0000000000000000000000000000000000000000000000000000000000000000
Solutions: 3
ZcashMiner:
proof-of-work found
prevHash: 0000000000000000000000000000000000000000000000000000000000000000
    hash: 52b433e36b254b661aac5b49cd3d7dd610c54bfa73cfc7fc8e1652974fa899d8
  target: 7fffff0000000000000000000000000000000000000000000000000000000000
duration: 70

real    1m10.422s
user    1m10.166s
sys     0m0.272s
```

(I added "return;" into standalone.cpp right after it prints "duration".)

```
[z@super src]$ gprof standaloneminer | head -15
Flat profile:

Each sample counts as 0.01 seconds.
  %   cumulative   self              self     total           
 time   seconds   seconds    calls   s/call   s/call  name    
 50.25     18.96    18.96                             blake2b_compress_ref
 29.31     30.01    11.06        3     3.69     3.69  void std::__introsort_loop<__gnu_cxx::__normal_iterator<TruncatedStepRow<13ul>*, std::vector<TruncatedStepRow<13ul>, std::allocator<TruncatedStepRow<13ul> > > >, long, __gnu_cxx::__ops::_Iter_comp_iter<CompareSR> >(__gnu_cxx::__normal_iterator<TruncatedStepRow<13ul>*, std::vector<TruncatedStepRow<13ul>, std::allocator<TruncatedStepRow<13ul> > > >, __gnu_cxx::__normal_iterator<TruncatedStepRow<13ul>*, std::vector<TruncatedStepRow<13ul>, std::allocator<TruncatedStepRow<13ul> > > >, long, __gnu_cxx::__ops::_Iter_comp_iter<CompareSR>)
  4.89     31.86     1.85 100673581     0.00     0.00  void std::__unguarded_linear_insert<__gnu_cxx::__normal_iterator<TruncatedStepRow<13ul>*, std::vector<TruncatedStepRow<13ul>, std::allocator<TruncatedStepRow<13ul> > > >, __gnu_cxx::__ops::_Val_comp_iter<CompareSR> >(__gnu_cxx::__normal_iterator<TruncatedStepRow<13ul>*, std::vector<TruncatedStepRow<13ul>, std::allocator<TruncatedStepRow<13ul> > > >, __gnu_cxx::__ops::_Val_comp_iter<CompareSR>)
  3.82     33.30     1.44        1     1.44    17.77  Equihash<96u, 3u>::OptimisedSolve(crypto_generichash_blake2b_state const&)
  2.55     34.26     0.96 67119159     0.00     0.00  TruncatedStepRow<13ul>::TruncatedStepRow<13ul>(TruncatedStepRow<13ul> const&, TruncatedStepRow<13ul> const&, unsigned long, unsigned long, int)
  1.78     34.93     0.67 33554432     0.00     0.00  StepRow<13ul>::StepRow(unsigned int, crypto_generichash_blake2b_state const&, unsigned int)
  1.74     35.58     0.66                             crypto_generichash_blake2b__final
  1.70     36.22     0.64 33562210     0.00     0.00  TruncatedStepRow<14ul>::TruncatedStepRow<13ul>(TruncatedStepRow<13ul> const&, TruncatedStepRow<13ul> const&, unsigned long, unsigned long, int)
  1.58     36.82     0.60       41     0.01     0.01  void std::__introsort_loop<__gnu_cxx::__normal_iterator<FullStepRow<38ul>*, std::vector<FullStepRow<38ul>, std::allocator<FullStepRow<38ul> > > >, long, __gnu_cxx::__ops::_Iter_comp_iter<CompareSR> >(__gnu_cxx::__normal_iterator<FullStepRow<38ul>*, std::vector<FullStepRow<38ul>, std::allocator<FullStepRow<38ul> > > >, __gnu_cxx::__normal_iterator<FullStepRow<38ul>*, std::vector<FullStepRow<38ul>, std::allocator<FullStepRow<38ul> > > >, long, __gnu_cxx::__ops::_Iter_comp_iter<CompareSR>)
  0.40     36.97     0.15        2     0.08     0.08  void std::vector<TruncatedStepRow<13ul>, std::allocator<TruncatedStepRow<13ul> > >::_M_range_insert<__gnu_cxx::__normal_iterator<TruncatedStepRow<13ul>*, std::vector<TruncatedStepRow<13ul>, std::allocator<TruncatedStepRow<13ul> > > > >(__gnu_cxx::__normal_iterator<TruncatedStepRow<13ul>*, std::vector<TruncatedStepRow<13ul>, std::allocator<TruncatedStepRow<13ul> > > >, __gnu_cxx::__normal_iterator<TruncatedStepRow<13ul>*, std::vector<TruncatedStepRow<13ul>, std::allocator<TruncatedStepRow<13ul> > > >, __gnu_cxx::__normal_iterator<TruncatedStepRow<13ul>*, std::vector<TruncatedStepRow<13ul>, std::allocator<TruncatedStepRow<13ul> > > >, std::forward_iterator_tag)
```

Time spent inside libc functions is not explicitly seen here, so my poor man's first attempt was actually better in that respect, but these profiling results do provide other insight.

**Comment by str4d** [comment]
Also see the memory benchmark runs on the CI server for massif output (e.g. http://ci.leastauthority.com:8010/builders/zcashperformance/builds/78/steps/memory%20solveequihash/logs/stdio ).

**Comment by solardiz** [comment]
> these profiling results do provide other insight

Hmm, but they only account for 37.72 seconds out of ~70 seconds. Maybe there are portions that I haven't rebuilt with `-pg` but where much time is spent as well, or it could be something else. Either way, the time percentages shown here are not of the total ~70 seconds, but are of 37.72 seconds.

**Comment by daira** [comment]
Well that's interesting. My impression was that an Equihash implementation without the hash-on-the-fly modification was supposed to spend most of its time sorting, not hashing.

**Comment by jl777** [comment]
having 100% of the code in user space will dramatically speed things up and also allow to get accurate benchmarks on all aspects. also what -O optimization level was used? -O2 can be significantly faster and all the performance characteristics will vary by incredible amounts if the active dataset fits into the L1/L2 cache

The problem with C++ is that it does things to help you automatically and there doesnt seem to be a way to control it directly, other than by not using any C++ construct that can trigger this "helpfulness"

when we are counting cycles and relying on data in the cache to stay there and not get overwritten, having some garbage collection interrupt everything right in the middle, well, it doesnt help. 

The actual performance can vary from what it should do theoretically by an order of magnitude, especially if external system calls are involved in the process

**Comment by solardiz** [comment]
> was supposed to spend most of its time sorting, not hashing.

19 out of 70 seconds isn't "most of its time", and this will probably be halved with the move from -ref to a SIMD implementation of BLAKE2b. That said, this is nevertheless very far from negligible.

**Comment by daira** [comment]
Re: the 37.72 vs ~70 second discrepancy, is the benchmark accidentally doing twice as many solution runs as it's supposed to? I ask because [this test in miner.cpp](https://github.com/zcash/zcash/blob/zc.v0.11.2.latest/src/miner.cpp#L543) causes the miner to do two solution runs in its inner loop.

**Comment by daira** [comment]
Oh, but of that treat were the problem then yous have two prints of "Running Equihash solver..."

**Comment by solardiz** [comment]
@jl777 For this build that I profiled I used `-O2 -momit-leaf-frame-pointer`, but I also kept `-fstack-protector-all -fPIE`, which might have some performance cost (typically tiny on x86_64, which this is, but possibly a few percent on 32-bit x86, which would be a concern for those builds). Also, profiling has some performance cost of its own.

**Comment by jl777** [comment]
as long as -O2 is on, then you will get 90% to 95% of optimal performance
half the time in system calls, doing who knows what
a quarter time hashing
a bit less time sorting

however, when 100% of the code is running without any system calls that add random delays, then all the performance ratios can change. if the current hashing is not fully local, then when that is fixed, sorting could well be the bottleneck

but if C++ is used, I dont think you can really control things at the level of granularity needed. Of course with OpenCL you get to specify the type of memory different variables go and that allows for really optimizing things to the theoretical max. 

Like automatic transmission vs. manual, former is great for minivans and grocery shopping... latter for race car driving

**Comment by daira** [comment]
Let's get rid of that inner loop. It may be contributing to the stale block problem, and it cannot help. (In bitcoin it was an optimisation to only update the chain after a significant number of hash trials, but the overhead of that is negligible in our case.)

**Comment by jl777** [comment]
definitely! when a hash takes 1 millionth the expected blocktime, it makes sense to do it occasionally, at one half the blocktime, it is totally different

**Comment by str4d** [comment]
@daira the inner loop is currently a no-op, only left there because a) at the time we didn't know if we'd need it again, and B) it made the diff from upstream smaller. But B) isn't a concern now because upstream dropped the miner entirely, so I'll remove it when refactoring.

**Comment by daira** [comment]
Oh wait, *is* the overhead negligible? We'd better check how long the chain update is actually taking, since that's included in the benchmark measurement iiuc.

**Comment by daira** [comment]
I think you're mistaken about the loop being a no-op. The ``break`` only occurs when the nonce is even.

**Comment by jl777** [comment]
compared to 60 seconds to find a solution set, any local blockchain query would be small and that would avoid wasting time working on a known stale branch

**Comment by str4d** [comment]
@daira oh whoops, you're right.

**Comment by daira** [comment]
However, it's not good benchmarking practice to include the chain update even if it is small.

**Comment by jl777** [comment]
#ifndef GOOD_BENCHMARKING_PRACTICE
// chain update
#endif

**Comment by daira** [comment]
I've just spotted another (minor) bug: if we find a proof of work then we don't increment the nonce before updating the chain, so if the chain hasn't actually changed then we will uselessly repeat the same solver run.

**Comment by jl777** [comment]
I know that theoretically any bit change in the nonce should get a random possible solution, however one of the benchmarks used for hashing functions is the cascade effect of changing one bit in the input on the output hash.

which tells me that in reality, changing just one bit isnt really getting a purely random new set of solutions and using a random set of nonces would seem to be to avoid getting stuck in any local minima without solutions. 

now incrementing the nonce does allow to deterministically mine, but if all miners are doing this, they are all doing the identical set of calculations. so on a global basis, this is quite redundant and wouldnt it mean that a faster machine will always win?

If each miner isnt mining from a different starting nonce, when the mempool has the same set of tx, it seems that odds for slow systems is 0%. To avoid it seems worth it to use /dev/urand for each nonce

but maybe there is some requirement for all nodes to calculate using the identical incrementing nonce?

**Comment by daira** [comment]
This is on the engineering meeting agenda in order to whether/when//how/who to implement a plain-C Equihash implementation.

**Comment by str4d** [comment]
> I've just spotted another bug: if we find a proof of work then we don't increment the nonce before updating the chain, so if the chain hasn't actually changed then we will uselessly repeat the same solver run.

That's not actually the case - even if the chain has not changed, the block time is updated, which changes the input to the Equihash solver.

**Comment by daira** [comment]
@jl777: Blake2b can safely be assumed to produce a fully pseudo-random output for any bit change in the input.

**Comment by daira** [comment]
@str4d You're right; still let's do the nonce increment unconditionally to simplify the code.

**Comment by jl777** [comment]
@daira, ok, even assuming blake2b is magically a perfect hashing function in this respect, the overlap of nonces across all miners (if that is the case) is not good.

you can start with a high entropy seed (the miner's pubkey) and use any sort of pseudorandom generator for the nonce. 

not sure if all nodes are mining the same nonces, so maybe it is a moot point

**Comment by str4d** [comment]
@jl777
> now incrementing the nonce does allow to deterministically mine, but if all miners are doing this, they are all doing the identical set of calculations. so on a global basis, this is quite redundant and wouldnt it mean that a faster machine will always win?

Generally no, because they are highly unlikely to have exactly-synchronised clocks, and thus the timestamps in their block headers will differ. But given the size of our nonce space, we can and will randomise the starting nonce anyway (as I have said several times in other places).

**Comment by jl777** [comment]
???
assuming that all clocks are within +/- 10 seconds, that means 10% of the nodes are overlapped. and in this day of ntp, most of the nodes could be on the exact same timestamp.

ok, randomize nonce solves this

**Comment by str4d** [comment]
(Also, another even more trivial reason there will not be overlap is because different miners will have different coinbase recipient addresses, which means they will have different Merkle tree roots, which means different Equihash inputs.)

**Comment by jl777** [comment]
aha! forgot about that, sorry for false alert, need more coffee

**Comment by daira** [comment]
@jl777 Duplication of nonces across miners isn't a problem, but duplication across threads is, and that is issue #1033.

**Comment by solardiz** [comment]
> half the time in system calls

@jl777 You're misreading this. While from my gprof run it is unclear where half the time is spent, it certainly is not in system calls. Per `time` output, system time is below 300ms out of 70s. Also, if it were syscalls called from functions being tracked by gprof, then their duration would be counted to the calling functions and would be seen in the gprof report. You're also making weird statements that C++ somehow uses system calls a lot. C++ uses library calls (in fact, plain C also sometimes does - just less commonly); whether those result in syscalls or (usually) not then depends on how the library is implemented. That said, and like I said before, I am also strongly in favor of moving to plain C for the performance-critical code. I just don't want us to exaggerate.

**Comment by jl777** [comment]
I dont want to exaggerate, I mean to say that C++ can be doing things in the background, not that it always does, just that it could. And using a library subjects you to whatever that library does, yes. so any C implementation would need to have C source for whatever library.

I did misunderstand the profile report though, it only seemed to add up to half the total time...
My assumption was that it was inside system calls that werent part of the profiling, but that is just an assumption. until all the code is profiled, all we can do is guess

also time output does not properly account for time spent in system calls in user space. Not sure what happens if a call ends up swapping to HDD and 10 milliseconds are burned. maybe that isnt counted anywhere, but certainly is significant.

So, where is half the time being spent?

**Comment by solardiz** [comment]
@jl777 There's no such thing as "system calls in user space". There are syscalls from userspace (and in fact only from it, as calls inside the kernel are not called that), which are calls into the kernel. "if a call ends up swapping to HDD and 10 milliseconds are burned" then a smart kernel and driver implementation may put that process/thread to sleep and let another one run on the same CPU core during those 10ms, if there is another runnable thread. In that case, the 10ms may not be counted towards the process' "system" time, but would be seen as part of its "real" time. On my profiling run, you can see that real=user+sys almost exactly, so this was not happening much.

I don't know where half the time is being spent. I could proceed to find out, but maybe it's more productive to work on that when we have C code, use SIMD BLAKE2b, etc. Per my earlier interrupt/continue experiment, it appeared that a lot of time, possibly over 50%, was spent in mem*() functions in libc. (Is this what you mean by "system calls in user space"? Those are library calls, not system calls.) If they were called from something that is profiled, I would expect them to be counted towards the calling functions, but maybe I'm wrong.

**Comment by jl777** [comment]
yes sorry i mean library calls, that are inside non-profiled libs, sometimes dynamically linked modules. which behave a lot like they are part of the system, even though it executes inside user space code.

so a specific system would behave different depending on environment variables and OUTSIDE the control of the miner, hence I used the term "system" improperly.

only way to get a full profile is to make sure all the code needed for miner is in source form part of the project, regardless of the language its in.

so we are saying the same thing, just in different ways. i definitely prefer getting things to C, then I can actually help with the code, not just high level things

**Comment by mcelrath** [comment]
It's worth noting that the NVidia compiler nvcc supports C++11.  It will be possible to GPU mine using CUDA on NVidia devices, no need for a C port.  (I think)

**Comment by daira** [comment]
@solardiz wrote:
> I don't know where half the time is being spent. I could proceed to find out, but maybe it's more productive to work on that when we have C code, use SIMD BLAKE2b, etc.

I disagree that we should block on having C code before trying to debug things like this. I don't know where the non-profiled time is going but I'm skeptical about the assumption that it has anything to do with C vs C++.

**Comment by daira** [comment]
@jl777 libc is profiled.

**Comment by jl777** [comment]
i agree perf is unlikely to be any C vs C++ issue
but the missing 50% is certainly half the time

**Comment by solardiz** [comment]
BTW, while the unaccounted-for half of the processing time is a profiling artifact, 2x memory allocation is for real. Somewhere half way through computation of a solution, memory allocation doubles, but this extra allocation is never made use of (not written to - mostly or at all). Known? Intentional?

**Comment by nathan-at-least** [comment]
Let's postpone the multi-threaded equihash implementation to a later milestone.

**Comment by Cryptoslave** [comment]
@str4d Beside multi-thread,  what do you think about using other n and k parameter such as k=10, n=198 which is 19.2 times faster as suggested by daira here [https://github.com/zcash/zcash/issues/1055#issuecomment-228622273]

**Comment by str4d** [comment]
@Cryptoslave it's a possibility, but note that the calculated speed increase is purely theoretical, and the actual speed would depend on the implementation (since those parameters are not supported currently). It would be nice if it did eventuate in practice, of course, but I think there are other avenues to explore first.

**Comment by str4d** [comment]
I've just run some cycle measurements on the testnet (`n = 96, k = 3`) using my new laptop (Intel(R) Core(TM) i7-6600U CPU @ 2.60GHz, 16GB RAM). Notes:

- No fancy tweaks to the CPU for measurement, so bear that in mind when interpreting the results.
- Only the first two rounds are shown separately; the final round is bundled into `Total cycles` for basic, and `Total for finding partials` for optimised. The cycles for eliminating partials in optimised are counted in its `Total cycles`.

`BasicSolve()`
--
Peak memory: ~1700MB VIRT, ~770MB RES
```
Filling 31903.17  Mcycles 
Round 1 41566.95  Mcycles 
Round 2 50784.83  Mcycles 
Total cycles 155328.35  Mcycles

Filling 31909.04  Mcycles 
Round 1 41312.96  Mcycles 
Round 2 50831.37  Mcycles 
Total cycles 154961.09  Mcycles

Filling 32321.06  Mcycles 
Round 1 41227.34  Mcycles 
Round 2 50952.42  Mcycles 
Total cycles 155225.59  Mcycles

Filling 31976.80  Mcycles 
Round 1 41367.01  Mcycles 
Round 2 50676.26  Mcycles 
Total cycles 154946.13  Mcycles
```

`OptimisedSolve()`
--
Peak memory: ~1400MB VIRT, ~480MB RES (ignoring the reallocation bug at the end that occurs on the CI server)
```
Filling 31763.10  Mcycles 
Round 1 28538.87  Mcycles 
Round 2 28098.97  Mcycles 
Total for finding partials 119000.92  Mcycles 
Total cycles 120806.53  Mcycles 

Filling 31936.40  Mcycles 
Round 1 28568.88  Mcycles 
Round 2 28451.17  Mcycles 
Total for finding partials 119325.88  Mcycles 
Total cycles 131578.01  Mcycles 

Filling 32322.82  Mcycles 
Round 1 28702.44  Mcycles 
Round 2 28300.63  Mcycles 
Total for finding partials 119741.52  Mcycles 
Total cycles 123265.55  Mcycles
```

Comparison to authors' implementation
--

For comparison, here is the output of `./equihash -n 96 -k 3 -S ####` for a few values. The post-filling memory usage (per htop) is about 5.8GB (both VIRT and RES), with peak about 12.1GB.

```
$ ./equihash -n 96 -k 3 -S 1337
N:	96 
K:	3 
SEED:  	0  	0  	0  	0 
Memory:		983040KiB
Testing nonce 2
Filling 79560.25  Mcycles 
Resolving 82266.24  Mcycles 
Resolving 94548.17  Mcycles 
Resolving 8849.15  Mcycles 
Time spent for n=96 k=3  983040 KiB: 265224.01  Mcycles 
Solution found:
 de985d  1ea9fb0  780cce  1f0581b  5d6486  bf7bef  10c905c  1467b83

$ ./equihash -n 96 -k 3 -S 13375
N:	96 
K:	3 
SEED:  	0  	0  	0  	0 
Memory:		983040KiB
Testing nonce 2
Filling 81182.07  Mcycles 
Resolving 76235.70  Mcycles 
Resolving 96613.61  Mcycles 
Resolving 8861.62  Mcycles 
Time spent for n=96 k=3  983040 KiB: 262893.20  Mcycles 
Solution found:
 de985d  1ea9fb0  780cce  1f0581b  5d6486  bf7bef  10c905c  1467b83
```

*plugged power in*

```
$ ./equihash -n 96 -k 3 -S 1337
N:	96 
K:	3 
SEED:  	0  	0  	0  	0 
Memory:		983040KiB
Testing nonce 2
Filling 79499.66  Mcycles 
Resolving 58224.51  Mcycles 
Resolving 93505.93  Mcycles 
Resolving 8660.31  Mcycles 
Time spent for n=96 k=3  983040 KiB: 239890.62  Mcycles 
Solution found:
 de985d  1ea9fb0  780cce  1f0581b  5d6486  bf7bef  10c905c  1467b83

$ ./equihash -n 96 -k 3 -S 13375
N:	96 
K:	3 
SEED:  	0  	0  	0  	0 
Memory:		983040KiB
Testing nonce 2
Filling 80529.18  Mcycles 
Resolving 57754.44  Mcycles 
Resolving 93765.72  Mcycles 
Resolving 8701.89  Mcycles 
Time spent for n=96 k=3  983040 KiB: 240751.41  Mcycles 
Solution found:
 de985d  1ea9fb0  780cce  1f0581b  5d6486  bf7bef  10c905c  1467b83
```

**Comment by str4d** [comment]
Split off the change in sorting algorithm to #1132.

**Comment by zookozcash** [comment]
This ticket is uncloseable because "optimize a thing" is a never-ending goal. Let's pick a limit, either a time-box after which we close this ticket and write down how far we got, or else a goal, like a certain CPU and RAM goal.

**Comment by Cryptoslave** [comment]
@str4d OK, but at least a choice of other n and k parameter has the possibility to be a radical change that may in one step have a far greater positive impact than many small optimisations. If it was me I would give it a go just because of big impact for I think smaller amount of work. Not much to loss by trying it and great gain is possible.

**Comment by nathan-at-least** [comment]
We're closing this ticket because it doesn't have a specific close criterion. We've split out specific optimiztions into subtickets.
